controllers/add_pods.go

```diff
@@ -75,217 +62,46 @@ func (a AddPods) Reconcile(r *FoundationDBClusterReconciler, context ctx.Context
 		return false, err
 	}
 
-	instanceIDs := make(map[string]map[int]bool)
+	instanceMap := make(map[string]FdbInstance, len(instances))
 	for _, instance := range instances {
-		instanceID := instance.GetInstanceID()
-		_, num, err := ParseInstanceID(instanceID)
-		if err != nil {
-			return false, err
-		}
-
-		class := instance.GetProcessClass()
-		if instanceIDs[class] == nil {
-			instanceIDs[class] = make(map[int]bool)
-		}
-
-		if instance.Pod != nil && instance.Pod.DeletionTimestamp != nil && !cluster.InstanceIsBeingRemoved(instanceID) {
-			return false, ReconciliationNotReadyError{message: "Cluster has pod that is pending deletion", retryable: true}
-		}
-
-		instanceIDs[class][num] = true
+		instanceMap[instance.GetInstanceID()] = instance
 	}
 
-	for _, processClass := range fdbtypes.ProcessClasses {
-		desiredCount := desiredCounts[processClass]
-		if desiredCount < 0 {
-			desiredCount = 0
-		}
-		newCount := desiredCount - currentCounts[processClass]
-		if newCount > 0 {
-			r.Recorder.Event(cluster, "Normal", "AddingProcesses", fmt.Sprintf("Adding %d %s processes", newCount, processClass))
-
-			pvcs := &corev1.PersistentVolumeClaimList{}
-			err = r.List(context, pvcs, getPodListOptions(cluster, processClass, "")...)
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		_, instanceExists := instanceMap[processGroup.ProcessGroupID]
+		if !instanceExists && !processGroup.Remove {
+			_, idNum, err := ParseInstanceID(processGroup.ProcessGroupID)
 			if err != nil {
 				return false, err
 			}
-			reusablePvcs := make(map[int]bool, len(pvcs.Items))
 
-			for index, pvc := range pvcs.Items {
-				ownedByCluster := false
-				for _, ownerReference := range pvc.OwnerReferences {
-					if ownerReference.UID == cluster.UID {
-						ownedByCluster = true
-						break
-					}
-				}
-
-				if ownedByCluster && pvc.ObjectMeta.DeletionTimestamp == nil {
-					instanceID := GetInstanceIDFromMeta(pvc.ObjectMeta)
-					if cluster.InstanceIsBeingRemoved(instanceID) {
-						continue
-					}
-
-					matchingInstances, err := r.PodLifecycleManager.GetInstances(
-						r, cluster, context,
-						getPodListOptions(cluster, processClass, instanceID)...,
-					)
-					if err != nil {
-						return false, err
-					}
-					if len(matchingInstances) == 0 {
-						reusablePvcs[index] = true
-					}
-				}
+			pod, err := GetPod(cluster, processGroup.ProcessClass, idNum)
+			if err != nil {
+				r.Recorder.Event(cluster, "Error", "GetPod", fmt.Sprintf("failed to get the PodSpec for %s/%d with error: %s", processGroup.ProcessClass, idNum, err))
```

Comment:
We might want to add the Pod name here (it might be contained in the error message)?

Reply:
I'm not sure the pod name will be relevant here, since the errors would come when building the pod spec, and we won't have an existing pod to reference when debugging the error.